block-layer/content-addressable-archives.md

```diff
@@ -30,51 +30,110 @@
 
 The CAR format (Content Addressable aRchives) can be used to store content addressable objects in the form of IPLD block data as a sequence of bytes; typically in a file with a `.car` filename extension.
 
-The CAR format is intended as a serialized representation of any IPLD DAG (graph) as the concatenation of its blocks, plus a header that describes the graphs in the file (via root CIDs). The requirement for the blocks in a CAR to form coherent DAGs is not strict, so the CAR format may also be used to store arbitrary IPLD blocks.
+The CAR format is intended as a serialized representation of any IPLD DAG (graph) as the concatenation of its blocks, plus a header that describes the graphs in the file (via root CIDs and selectors). The requirement for the blocks in a CAR to form coherent DAGs is not strict, so the CAR format may also be used to store arbitrary IPLD blocks.
 
 In addition to the binary block data, storage overhead for the CAR format consists of:
 
- * A header block encoded as [DAG-CBOR](codecs/dag-cbor.md) containing the format version and an array of root CIDs
- * A CID for each block preceding its binary data
- * A compressed integer prefixing each block (including the header block) indicating the total length of that block, including the length of the encoded CID
+ * A compressed integer
+ * A "version" block of byte-length described by the preceding compressed integer
+ * A series of CID-prefixed blocks:
+   * A compressed integer prefixing each block indicating the total byte-length of that block, including the length of the encoded CID
```

Comment:
I wonder if it would be useful to have an int for the CID length and an int for the content data length, separately.

Figuring out how long a CID is... is a surprisingly nontrivial operation.  https://github.com/ipfs/go-cid/pull/95/files recently added support for this, but involves at least three different varint parses (spread across more than just that one repo!) -- and in the meanwhile, you're passing around an overly large byte slice.  (How much too large?  I dunno.  I guess you just take some bytes and wing it...?)

So, considering all that... while technically, sure, it's "defined" to just put the CID at the front of the content bytes and just report the total size... I'd advocate for an int to state the CID size separately because it's more antifragile, and simpler, and might even result in more efficient processing since it'll let code spend more time operating with known amounts of bytes.

Reply:
I haven’t seen a CBOR parser decode this as fast as we could parse the int based encoding so I’m worried about moving in that direction.

I do like the idea of separating the ints. Right now you’ve basically gotta parse the entire CID in order to work through the file. If you have a specific set of CID’s you’re scanning for you could do much faster binary comparison as you scan through the file and avoid full CID parsing.

Reply:
Thinking this through a bit more combining with https://github.com/ipld/specs/pull/248#issuecomment-597057196 :

Having v2 being a uniform stream of `{cidlen}{cid}{payloadlen}{payload}` where `{cid}` ( position 2 ) has a special codec signifying `car header` as opposed to just `dag-cbor` would solve all your symmetry issues and be super-fast to check for well-formedness...